allocations/tasks.py

```diff
@@ -1,18 +1,137 @@
+    charge_code = alloc.project.charge_code
+    email = alloc.project.pi.email
+    expiration_date = alloc.expiration_date.date()
+
+    # Edge cases
+    if not expiration_date:
+        LOG.warning(f"Allocation {alloc.id} has no expiration date.")
+        return None
+
+    if not email:
+        LOG.warning(
+            f"PI for project {charge_code} has no email; "
+            "cannot send expiration warning"
+        )
+        return None
+
+    time_until_expiration = expiration_date - today
+
+    # Little string to describe the expiration to make the email look nicer
+    if expiration_date == today:
+        time_description = "today"
+    else:
+        # "in 1 day" | "in <n> days"
+        time_description = (
+            f"in {time_until_expiration.days} "
+            f"day{'s' if time_until_expiration.days != 1 else ''}"
+        )
+
+    docs_url = (
+        "https://chameleoncloud.readthedocs.io/en/latest/user/project.html"
+        "#recharge-or-extend-your-allocation "
+    )
+    email_body = f"""
+            <p>
+                The allocation for project <b>{charge_code}</b> 
+                will expire <b>{time_description}.</b> See our
+                <a href={docs_url}>Documentation</a>
+                on how to recharge or extend your allocation.
+            </p>
+            """
+
+    mail_sent = None
+    try:
+        mail_sent = send_mail(
+            subject=f"NOTICE: Your allocation for project {charge_code} "
+            f"expires {time_description}!",
+            from_email=settings.DEFAULT_FROM_EMAIL,
+            recipient_list=[email],
+            message=strip_tags(" ".join(email_body.split()).strip()),
+            html_message=email_body,
+        )
+    except Exception:
+        pass
+
+    return mail_sent
+
+
+def warn_user_for_expiring_allocation():
+    """
+    Sends an email to users when their allocation is within one month of expiring
+    """
+    now = datetime.now(timezone.utc)
+    today = now.date()
+    days_this_month = calendar.monthrange(today.year, today.month)[1]
+    target_expiration_date = today + timedelta(days=days_this_month)
+
+    # Find allocations that are expiring between today and 1 month,
+    # and have not been warned of their impending doom
+    expiring_allocations = Allocation.objects.filter(
+        status='active',
+        expiration_date__lte=target_expiration_date,
+        expiration_date__gte=today,
+        expiration_warning_issued__isnull=True,
+    )
+
+    emails_sent = 0
+    for alloc in expiring_allocations:
+        mail_sent = _send_expiration_warning_mail(alloc, today)
+        charge_code = alloc.project.charge_code
+
+        # If we successfully sent mail, log it in the database
+        if mail_sent:
+            emails_sent += 1
+            LOG.info(
+                f"Warned PI about allocation {alloc.id} "
+                f"expiring on {alloc.expiration_date}"
+            )
+            if alloc.expiration_warning_issued:
```

Comment:
Instead of doing this, I think an `assert alloc.expiration_warning_issued is None` is warranted in the _send_expiration_warning_mail. Asserts are good for things that the code should already be ensuring, but you're just making sure. Asserts fire only if a bad code change was introduced; they're a good circuit breaker before something critical gets called w/ bad input.

Reply:
Good call!